### PR TITLE
Add option to skip self loops for acceptance functions

### PR DESCRIPTION
**Pull request by matthewkol186**
# Description
As discussed with @pjrule , I am putting in a PR that allows the user to easily specify whether a failure to accept should constitute a self-loop (which counts towards the total # of steps) or whether self-loops should be disallowed (i.e., new proposals are generated until one satisfies an acceptance loop, at which point the step counter is incremented). Note that this feature has the potential to create infinite loops, if the user specifies `no_self_loops=true`. 

# Usage
```
chain_data = recom_chain(graph, partition, pop_constraint, num_steps, scores, acceptance_fn=f, no_self_loops=true)
```

**PROMISE**: I will update documentation before merging this code!

### CODE REVIEW

File: src/recom.jl [hunk 1]
```diff
@@ -223,6 +228,10 @@ function recom_chain(graph::BaseGraph,
         if custom_acceptance && !satisfies_acceptance_fn(partition, acceptance_fn)
             # go back to the previous partition
             partition = partition.parent
+            # if user specifies this behavior, we do not increment the steps
+            # taken if the acceptance function fails. BEWARE - this can create
```
Comment (pjrule) [inline src/recom.jl:232]: This `BEWARE!` should probably be in the docstring for both chains.

File: test/recom.jl [hunk 2]
```diff
@@ -1,6 +1,11 @@
 @testset "Recom tests" begin
     graph = BaseGraph(square_grid_filepath, "population", "assignment")
 
+    function accept_on_third_try()
```
Comment (pjrule) [inline test/recom.jl:4]: clever! :)

**Review by pjrule — APPROVED** [review]
up to any minor changes in documentation, this lgtm.

File: test/recom.jl (hunk 2 above)
Comment (matthewkol186) [reply test/recom.jl:4]: hehe i have my moments 😏

File: src/recom.jl (hunk 1 above)
Comment (matthewkol186) [reply src/recom.jl]: Done!

**Pull request merged**